query/src/test/java/com/connexta/replication/query/QueryServiceTest.java

```diff
@@ -0,0 +1,165 @@
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class QueryServiceTest {
+
+  private static final long PERIOD = 60L;
+
+  private static final String SITE_ID = "siteid";
+
+  @Mock private Site site;
+
+  @Mock private QueryServiceTools queryServiceTools;
+
+  @Mock private ScheduledExecutorService executor;
+
+  @Mock private SiteQueue siteQueue;
+
+  @Mock private NodeAdapter adapter;
+
+  private QueryService queryService;
+
+  @Before
+  public void setUp() throws Exception {
+    when(queryServiceTools.getQueueFor(site)).thenReturn(siteQueue);
+    when(queryServiceTools.getAdapterFor(site)).thenReturn(adapter);
+    when(queryServiceTools.getPeriod()).thenReturn(PERIOD);
+    queryService = new QueryService(site, queryServiceTools, executor);
+  }
+
+  @Test
+  public void update() {
+    Site site2 = mock(Site.class);
+    queryService.update(site2);
+    verify(queryServiceTools).getAdapterFor(site2);
+  }
+
+  @Test
+  public void start() {
+    ScheduledFuture future = mock(ScheduledFuture.class);
+    when(executor.scheduleAtFixedRate(
+            any(Runnable.class), eq(0L), eq(PERIOD), eq(TimeUnit.SECONDS)))
+        .thenReturn(future);
+    queryService.start();
+    verify(executor)
+        .scheduleAtFixedRate(any(Runnable.class), eq(0L), eq(PERIOD), eq(TimeUnit.SECONDS));
+    assertThat(queryService.future, notNullValue());
+  }
+
+  @Test
+  public void query() throws Exception {
+    ArgumentCaptor<QueryRequest> requestArgumentCaptor =
+        ArgumentCaptor.forClass(QueryRequest.class);
+    // #query
+    when(adapter.isAvailable()).thenReturn(true);
+    Filter filter = mock(Filter.class);
+    when(queryServiceTools.activeFiltersFor(site)).thenReturn(Stream.of(filter));
+    FilterIndex filterIndex = mock(FilterIndex.class);
+    when(queryServiceTools.getOrCreateFilterIndex(filter)).thenReturn(filterIndex);
+
+    // #getNewMetadataForFilter
+    when(filterIndex.getModifiedSince()).thenReturn(Optional.empty());
+    Metadata metadata = mock(Metadata.class);
+    QueryResponse queryResponse = mock(QueryResponse.class);
+    when(adapter.query(any(QueryRequest.class))).thenReturn(queryResponse);
+    when(queryResponse.getMetadata()).thenReturn(List.of(metadata));
+
+    // #metadataToTask
+    when(filter.getPriority()).thenReturn((byte) 0);
+    when(metadata.getResourceSize()).thenReturn(0L);
+    Date lastModified = new Date();
+    when(metadata.getMetadataModified()).thenReturn(lastModified);
+
+    queryService.query();
+
+    verify(adapter).query(requestArgumentCaptor.capture());
+    assertThat(requestArgumentCaptor.getValue().getModifiedAfter(), is(nullValue()));
+    verify(siteQueue).put(any(TaskInfoImpl.class));
```

Comment:
❓ How do we know if the task created is the right one? Contains the right resource info and metadata info, the right priority, ...
We should have test cases that tests a few combination of those.